### ISSUE TITLE
enhance load of certificates

### ISSUE THREAD

**Issue by petrovr** [ronf/asyncssh#228]
Sometime X.509 certificate could be stored into file with extra information for instance result of command: openssl x509 ... -text.

It seems to me such certificate cannot be loaded as trusted in asyncssh. 
It is valid for openssl as it ignores "extra"  data before and after "certificate marks".
Note used marks:
(1) '-----BEGIN CERTIFICATE-----' with '-----END CERTIFICATE-----' and
(2) '-----BEGIN TRUSTED CERTIFICATE-----'  with  '-----END TRUSTED CERTIFICATE-----'

There is no request to consider (2) differently then (1). For instance OnenSSL consider trusted those available in 'store' (loaded from files and etc.) and un-trusted certificates received from wire.

**Comment by ronf** [comment]
Thanks for the report!

I was not familiar with the OpenSSH proprietary extension of "BEGIN TRUSTED CERTIFICATE" before this. Right now, AsyncSSH will ignore lines not matching the expected PEM type requested ("CERTIFICATE" in this case), but it also does a check in the case of certificates that the header is _exactly_ "BEGIN CERTIFICATE" without any other name in between (unlike public/private keys that can have a key type listed in some cases).

If you'd like to experiment with this, here's a quick patch to allow "TRUSTED CERTIFICATE" PEM blocks to load:

```patch
index 672568a..b424265 100644
--- a/asyncssh/public_key.py
+++ b/asyncssh/public_key.py
@@ -2315,7 +2315,7 @@ def _decode_pem_certificate(lines):
 
     pem_name, _, data, end = _decode_pem(lines, b'CERTIFICATE')
 
-    if pem_name:
+    if pem_name and pem_name != b'TRUSTED':
         raise KeyImportError('Invalid PEM certificate')
 
     return SSHX509Certificate.construct(data), end
```

Note that this doesn't attempt to parse any of the additional information. So, any addtrust/addreject rules associated with the certificate will be ignored when using it.

Also, I can't be 100% sure that the PyCA X.509 certificate support will always ignore the additional trust information which appears to be appended to the end of the base64 DER-encoded data within the PEM header/trailer. In a quick test, it seems ok, but I'd need to do more thorough testing to be sure everything works properly with these new certificates.

**Comment by petrovr** [comment]
It is OpenSSL not OpenSSH.

Enhancement is more about to use data only between "markers", i.e. to load files result from following command:
$ openssl x509 -in FILE ...  -text {-fingerprint}


About "trusted" . Not so important marker - even openssl "ignore" this extra information. I mean that if x.509 PEM file is created with  ... -addtrust .. -addreject ... -setalias and etc. then on read openssl x509 .. will output only normal information. This "extra" is part of encoded output only if is specified explicitely option -trustout.

**Comment by ronf** [comment]
> It is OpenSSL not OpenSSH.

Sorry - I knew it was OpenSSL, but apparently my fingers still typed "OpenSSH". I'll fix that!

Regarding only reading data between markers, the difficulty is that the read_certificate() function supports more than just PEM files. It uses the first few bytes of the file to detect whether the file is DER, PEM, or RFC 4716. If the file doesn't start with '-----' or '---- ', it assumes the file is DER format. As long as it started with ANY kind of PEM header, though, it will properly skip over any lines in the file (even if they aren't PEM data) until it gets to the section it is looking for. Since DER format data typically starts with a SEQUENCE byte ('\x30', which is an ASCII '0'), I could look for that instead, but there'd still be a chance of misidentification if the file was actually PEM data but it happened to start with an ASCII '0' on the first line.

Regarding ignoring the "TRUSTED" header, that's what the above patch will do, but I found that the current code will end up storing the extra trust data when it reads in the certificate, and later when it outputs the certificate in SSH format during a key exchange, that extra data may cause trouble. That's why I suggested additional testing might be needed.

It should be possible to parse the decoded base64 data to find the end of the regular certificate and strip off the extra trust data so that it never appears in the SSH handshake, but that will require additional code.

**Comment by ronf** [comment]
Ok - I think the following should properly ignore the data OpenSSL adds to "trusted" certificates:

```patch
index 672568a..424fca4 100644
--- a/asyncssh/public_key.py
+++ b/asyncssh/public_key.py
@@ -2315,7 +2315,14 @@ def _decode_pem_certificate(lines):
 
     pem_name, _, data, end = _decode_pem(lines, b'CERTIFICATE')
 
-    if pem_name:
+    if pem_name == b'TRUSTED':
+        # Strip off OpenSSL trust information
+        try:
+            _, end = der_decode(data, partial_ok=True)
+            data = data[:end]
+        except ASN1DecodeError:
+            raise KeyImportError('Invalid PEM trusted certificate') from None
+    elif pem_name:
         raise KeyImportError('Invalid PEM certificate')
 
     return SSHX509Certificate.construct(data), end
```

I'll also need to add unit tests for this new code path, but you're welcome to try the above in the meantime.

**Comment by ronf** [comment]
The above change is now available in the "develop" branch as commit fe9648e.

**Comment by petrovr** [comment]
Hi Rob,

It seems to me you implementation exceed requirement.
The projects that use OpenSSL store ( page https://roumenpetrov.info/domino_CA/ is not up to date ) are compatible with functionality described in X509_LOOKUP_hash_dir(3) .
Function X509_LOOKUP_file(3) and X509_LOOKUP_hash_dir(3) expect PEM format. So for load from file PEM only suitable format is PEM (DER does not support multiple certificates). 
Load from directory PEM is default. It seem to me you implementation exceed requirement with support for DER format. This is fine but not used in OpenSSL based programs. As you pint above check first byte could be used there.

About you commit (sorry for late replay) I would like to distinguish between (1) trusted certificate and (2) certificates with "trusted" (extra) information. From this point of view message is ambiguous.
If load succeed certificate will be considered as "trusted" right? If I understand properly issues is this "extra" OpenSSL specific information. Let me quote from OpenSSL manual "d2i_X509_AUX() is similar to d2i_X509() but the input is expected to consist of an X509 certificate followed by auxiliary trust information".
Please could you correct message. No idea how, something like "Certificate in PEM format with  auxiliary trust information is not supported".
Actually "trust information" is applicable for DER but proposed exception is only for PEM format.

**Comment by ronf** [comment]
> The projects that use OpenSSL store ( page https://roumenpetrov.info/domino_CA/ is not up to date ) are compatible with functionality described in X509_LOOKUP_hash_dir(3). Function X509_LOOKUP_file(3) and X509_LOOKUP_hash_dir(3) expect PEM format. So for load from file PEM only suitable format is PEM (DER does not support multiple certificates).
> 
> Load from directory PEM is default. It seem to me you implementation exceed requirement with support for DER format. This is fine but not used in OpenSSL based programs. As you pint above check first byte could be used there.

You're probably right that in the specific case of handling X.509 certificates I could have gotten away without supporting DER format for hash directories. However, the import_certificate() code in AsyncSSH is used to read both X.509 certificates and OpenSSH certificates, each in multiple formats. Also, unlike OpenSSL which has separate functions for reading DER vs. PEM, I allow the same function to be used for both, and I also support multiple certificates being in a single file when reading a certificate list in all formats (DER, PEM, OpenSSH, and RFC 4716).

I may have an alternate way of doing the format recognition that will allow arbitrary text to be skipped even at the beginning of a file. I'd just need to re-do the format detection after each certificate, making sure that lines that don't turn out to be a valid OpenSSH format certificate and aren't the beginning of a DER format certificate get properly skipped. I'll look into this further.

> About you commit (sorry for late replay) I would like to distinguish between (1) trusted certificate and (2) certificates with "trusted" (extra) information. From this point of view message is ambiguous.
> If load succeed certificate will be considered as "trusted" right? If I understand properly issues is this "extra" OpenSSL specific information. Let me quote from OpenSSL manual "d2i_X509_AUX() is similar to d2i_X509() but the input is expected to consist of an X509 certificate followed by auxiliary trust information".

My understanding of the OpenSSL functions you mention here is as follows:

- If you do the parsing of PEM and base64 decoding yourself, you should call d2i_X509_AUX() whenever the PEM header says "BEGIN TRUSTED CERTIFICATE", passing in the DER format binary data you get back from the base64 decode.
- Otherwise, if the PEM header says "BEGIN CERTIFICATE", you should call d2i_X509(), because the extra "trust" information is only present in the case where "BEGIN TRUSTED CERTIFICATE" appears and attempting to parse this extra trust information on regular certificates will fail.
- Regarding of which PEM header is present, there's no difference in which kind of certificates can be loaded as "trusted". That has more to do with how you construct the X.509 trust store, and what specific certificates you load into it and not which type of certificate ("TRUSTED" in the PEM header or not) is loaded.

In the case of using AsyncSSH with X.509 certificates, a peer certificate will only be trusted when both of the following are true:

- It's possible to build a valid and complete certificate chain from the peer's end-entity certificate up to some self-signed root certificate, with every step along the way having issuer names match subject names of the next certificate in the chain, and other information like validity dates and usage information all being set properly.
- The root certificate at the top of the certificate chain was one of the certificates loaded as "trusted". This can be done by specifying the certificates explicitly in the x509_trusted_certs or by placing the certificates in a hash dir with the path listed in x509_trusted_cert_paths. It can also be done by putting the certificates to trust in the default locations that AsyncSSH looks for (documented at https://asyncssh.readthedocs.io/en/latest/api.html#sshclientconnectionoptions). As noted there, the certificates can also be converted to OpenSSH format and included in the known_hosts file.

When intermediate certificate authorities are involved, it is expected that the peer will provide all the necessary intermediate CA certificates to connect the end-entity certificate with one of the trusted roots. However, if necessary, you can include intermediate CA certs in your x509_trusted_certs. This will NOT implicitly trust that intermediate CA, though. It will only make that certificate available to be used to complete a certificate chain between the peer's end-entity certificate and one of the self-signed trusted root certificates you have included.

> Please could you correct message. No idea how, something like "Certificate in PEM format with auxiliary trust information is not supported". Actually "trust information" is applicable for DER but proposed exception is only for PEM format.

This would not be correct. With the change I've made here, certificates in PEM format with auxiliary trust information ARE supported. This auxiliary information is ignored, though, meaning that none of the requested additional enforcements on trusted/untrusted purposes listed will be respected. If you wish to limit purposes, though, AsyncSSH already provides a different way to do that, using the x509_purposes argument.

Does this help to explain things better?

**Comment by ronf** [comment]
I just checked in commit 0c07f27 in the "develop" branch, improving the flexibility of the parser to support arbitrary text before valid key/certificate data. As part of this change, the parser can now also handle mixing multiple key formats in a file when reading a list of keys/certificates, with arbitrary text before, between, or after them.

**Comment by ronf** [comment]
These enhancements are now available in the AsyncSSH 1.18.0 release!
